sw/device/lib/crypto/drivers/hmac.c

```diff
@@ -8,148 +8,420 @@
 
-  for (; len >= sizeof(uint32_t); len -= sizeof(uint32_t)) {
-    uint32_t data_aligned = read_32(data);
-    abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_MSG_FIFO_REG_OFFSET,
-                     data_aligned);
-    data += sizeof(uint32_t);
+  // We need to keep `sha_en` low until context is restored, see #23014.
+  ctx->cfg_reg = bitfield_bit32_write(ctx->cfg_reg, HMAC_CFG_SHA_EN_BIT, false);
+
+  // Default value for `hmac_en` is false, HMAC calls set it to true below.
+  ctx->cfg_reg =
+      bitfield_bit32_write(ctx->cfg_reg, HMAC_CFG_HMAC_EN_BIT, false);
+
+  switch (hmac_mode) {
+    case kHmacModeHmac256:
+      ctx->cfg_reg =
+          bitfield_bit32_write(ctx->cfg_reg, HMAC_CFG_HMAC_EN_BIT, true);
+      OT_FALLTHROUGH_INTENDED;
+    case kHmacModeSha256:
+      ctx->cfg_reg =
+          bitfield_field32_write(ctx->cfg_reg, HMAC_CFG_DIGEST_SIZE_FIELD,
+                                 HMAC_CFG_DIGEST_SIZE_VALUE_SHA2_256);
+      // The unit for `ctx->msg_block_len` is bytes.
+      ctx->msg_block_len = kHmacSha256InternalBlockBytes;
+      // The unit for `ctx->digest_len` is bytes.
+      ctx->digest_len = kHmacSha256DigestBytes;
+      break;
+    case kHmacModeHmac384:
+      ctx->cfg_reg =
+          bitfield_bit32_write(ctx->cfg_reg, HMAC_CFG_HMAC_EN_BIT, true);
+      OT_FALLTHROUGH_INTENDED;
+    case kHmacModeSha384:
+      ctx->cfg_reg =
+          bitfield_field32_write(ctx->cfg_reg, HMAC_CFG_DIGEST_SIZE_FIELD,
+                                 HMAC_CFG_DIGEST_SIZE_VALUE_SHA2_384);
+      // The unit for `ctx->msg_block_len` is bytes.
+      ctx->msg_block_len = kHmacSha384InternalBlockBytes;
+      // The unit for `ctx->digest_len` is bytes.
+      ctx->digest_len = kHmacSha384DigestBytes;
+      break;
+    case kHmacModeHmac512:
+      ctx->cfg_reg =
+          bitfield_bit32_write(ctx->cfg_reg, HMAC_CFG_HMAC_EN_BIT, true);
+      OT_FALLTHROUGH_INTENDED;
+    case kHmacModeSha512:
+      ctx->cfg_reg =
+          bitfield_field32_write(ctx->cfg_reg, HMAC_CFG_DIGEST_SIZE_FIELD,
+                                 HMAC_CFG_DIGEST_SIZE_VALUE_SHA2_512);
+      // The unit for `ctx->msg_block_len` is bytes.
+      ctx->msg_block_len = kHmacSha512InternalBlockBytes;
+      // The unit for `ctx->digest_len` is bytes.
+      ctx->digest_len = kHmacSha512DigestBytes;
+      break;
+    default:
+      return OTCRYPTO_BAD_ARGS;
   }
 
-  // Handle non-32bit aligned bytes at the end of the buffer.
-  for (; len != 0; --len) {
-    abs_mmio_write8(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_MSG_FIFO_REG_OFFSET,
-                    *data++);
+  if (hmac_mode == kHmacModeHmac256 || hmac_mode == kHmacModeHmac384 ||
+      hmac_mode == kHmacModeHmac512) {
+    if (key == NULL) {
+      return OTCRYPTO_BAD_ARGS;
+    }
+    ctx->cfg_reg =
+        bitfield_bit32_write(ctx->cfg_reg, HMAC_CFG_HMAC_EN_BIT, true);
+    // Key values supported natively by HW are {128, 256, 384, 512, 1024}.
+    switch (key->len) {
+      case 128 / 8:
```

Comment:
Do you have any enum definitions in the hmac.h file you can use here?

Reply:
We do not have anything coming from the autogenerated header but I will add the following to `hmac.h`, so that we do not have these naked values:
```
enum {
  /* 128 bit key. */
  kHmacKey128Bytes = 128 / 8,
  /* 256 bit key. */
  kHmacKey256Bytes = 256 / 8,
  /* 384 bit key. */
  kHmacKey384Bytes = 384 / 8,
  /* 512 bit key. */
  kHmacKey512Bytes = 512 / 8,
  /* 1024 bit key. */
  kHmacKey1024Bytes = 1024 / 8,
};
```